// graphile-build/node8plus/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MutationPayloadQueryPlugin = exports.ClientMutationIdDescriptionPlugin = exports.MutationPlugin = exports.QueryPlugin = exports.NodePlugin = exports.StandardTypesPlugin = exports.defaultPlugins = exports.buildSchema = exports.getBuilder = undefined;

var _util = require("util");

var _util2 = _interopRequireDefault(_util);

var _SchemaBuilder = require("./SchemaBuilder");

var _SchemaBuilder2 = _interopRequireDefault(_SchemaBuilder);

var _plugins = require("./plugins");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const getBuilder = exports.getBuilder = async (plugins, options = {}) => {
  const builder = new _SchemaBuilder2.default();
  for (const plugin of plugins) {
    if (typeof plugin !== "function") {
      throw new Error("Expected a list of plugin functions, instead list contained a non-function: " + _util2.default.inspect(plugin));
    }
    builder._setPluginName(plugin.displayName || plugin.name);
    await plugin(builder, options);
    builder._setPluginName(null);
  }
  return builder;
};

const buildSchema = exports.buildSchema = async (plugins, options = {}) => {
  const builder = await getBuilder(plugins, options);
  return builder.buildSchema();
};

const defaultPlugins = exports.defaultPlugins = [_plugins.StandardTypesPlugin, _plugins.NodePlugin, _plugins.QueryPlugin, _plugins.MutationPlugin, _plugins.ClientMutationIdDescriptionPlugin, _plugins.MutationPayloadQueryPlugin];

exports.StandardTypesPlugin = _plugins.StandardTypesPlugin;
exports.NodePlugin = _plugins.NodePlugin;
exports.QueryPlugin = _plugins.QueryPlugin;
exports.MutationPlugin = _plugins.MutationPlugin;
exports.ClientMutationIdDescriptionPlugin = _plugins.ClientMutationIdDescriptionPlugin;
exports.MutationPayloadQueryPlugin = _plugins.MutationPayloadQueryPlugin;
