// graphile-build-pg/node8plus/plugins/PgQueryProceduresPlugin.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _makeProcField = require("./makeProcField");

var _makeProcField2 = _interopRequireDefault(_makeProcField);

var _debug = require("debug");

var _debug2 = _interopRequireDefault(_debug);

var _chalk = require("chalk");

var _chalk2 = _interopRequireDefault(_chalk);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const debugWarn = (0, _debug2.default)("graphile-build-pg:warn");

exports.default = function PgQueryProceduresPlugin(builder) {
  builder.hook("GraphQLObjectType:fields", (fields, build, { scope: { isRootQuery }, fieldWithHooks }) => {
    if (!isRootQuery) {
      return fields;
    }
    const {
      extend,
      pgInflection: inflection,
      pgIntrospectionResultsByKind: introspectionResultsByKind
    } = build;
    return extend(fields, introspectionResultsByKind.procedure.filter(proc => proc.isStable).filter(proc => !!proc.namespace).reduce((memo, proc) => {
      /*
      proc =
        { kind: 'procedure',
          name: 'integration_webhook_secret',
          description: null,
          namespaceId: '6484381',
          isStrict: false,
          returnsSet: false,
          isStable: true,
          returnTypeId: '2950',
          argTypeIds: [ '6484569' ],
          argNames: [ 'integration' ],
          argDefaultsNum: 0 }
      */
      const argTypes = proc.argTypeIds.map(typeId => introspectionResultsByKind.typeById[typeId]);
      if (argTypes.some(type => type.type === "c" && type.class && type.class.isSelectable)) {
        // Selects a table, ignore!
        return memo;
      }
      const firstArgType = argTypes[0];
      if (firstArgType && firstArgType.type === "c" && firstArgType.class && firstArgType.namespaceId === proc.namespaceId && proc.name.startsWith(`${firstArgType.name}_`)) {
        // It's a computed field, skip
        return memo;
      }

      const fieldName = inflection.functionName(proc.name, proc.namespace.name);
      try {
        memo[fieldName] = (0, _makeProcField2.default)(fieldName, proc, build, {
          fieldWithHooks
        });
      } catch (e) {
        // eslint-disable-next-line no-console
        console.warn(_chalk2.default.bold.yellow(`Failed to add function '${proc.namespace.name}.${proc.name}'; run with 'DEBUG="graphile-build-pg:warn"' to view the error`));
        debugWarn(e);
      }
      return memo;
    }, {}), `Adding query procedures to root Query type`);
  });
};
